test/jdk/java/nio/channels/FileChannel/ChannelInputStreamTest.java

```diff
@@ -0,0 +1,180 @@
+ * accompanied this code).
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+
+/* @test
+ * @bug 8361495
+ * @summary Test for AsynchronousCloseException from uninterruptible FileChannel
+ * @run junit ChannelInputStreamTest
+ */
+
+import java.io.Closeable;
+import java.io.InputStream;
+import java.io.IOException;
+import java.io.OutputStream;
+import java.nio.channels.AsynchronousCloseException;
+import java.nio.channels.ClosedChannelException;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.util.Arrays;
+import java.util.concurrent.ConcurrentLinkedQueue;
+
+import org.junit.jupiter.api.Test;
+
+public class ChannelInputStreamTest {
+    private static Thread startCloseThread(ConcurrentLinkedQueue<Object> q) {
+        return Thread.ofPlatform().start(() -> {
+            do {
+                Object obj;
+                if ((obj = q.poll()) != null) {
+                    if (obj instanceof Closeable c) {
+                        try {
+                            c.close();
+                        } catch (IOException ignored) {
+                        }
+                    } else {
+                        break;
+                    }
+                }
+            } while (true);
+        });
+    }
+
+    @Test
+    public void available() throws IOException {
+        var close = new ConcurrentLinkedQueue<Object>();
+        Thread closeThread = startCloseThread(close);
+
+        Path path = Files.createTempFile(Path.of("."), "foo", "bar");
+        path.toFile().deleteOnExit();
+
+        do {
+            InputStream in;
+            try {
+                in = Files.newInputStream(path);
+            } catch (IOException ignored) {
+                continue;
+            }
+            close.offer(in);
+            int available = 0;
+            try {
+                available = in.available();
+            } catch (AsynchronousCloseException ace) {
+                System.out.println("AsynchronousCloseException caught");
+                close.offer(new Object());
+                break;
+            } catch (ClosedChannelException ignored) {
+                continue;
+            } catch (Throwable t) {
+                close.offer(new Object());
+                throw new RuntimeException("Unexpected error", t);
+            }
+            if (available < 0) {
+                close.offer(new Object());
+                throw new RuntimeException("FAILED: available < 0");
+            }
+        } while (true);
+
+        try {
+            closeThread.join();
+        } catch (InterruptedException ignored) {
+        }
```

Comment:
I assume we shouldn't catch InterruptedExcepiton here, this could only happen if at test timeout and jtreg is attempted to cleanup.